test/extensions/quic_listeners/quiche/platform/envoy_quic_clock_test.cc

```diff
@@ -55,5 +54,25 @@ TEST(EnvoyQuicClockTest, TestMonotonicityWithReadTimeSystem) {
   }
 }
 
+TEST(EnvoyQuicClockTest, ApproximateNow) {
+  Event::SimulatedTimeSystemHelper time_system;
```

Comment:
Can you add another test to verify onPrepare() can update approximate_now? Such test might expose data race as you encountered before. You might want to access approximate_now via Dispatcher::post(). And as a sanity check, it would be better to add an ASSERT(dispatcher_.isThreadSafe()) wherever approximate_now_ is accessed.

Reply:
Done.  Turns out I need to post a task in order for dispacher->run() to call the prepare callback.